src/setup-elixir.js

```diff
@@ -1,97 +1,356 @@
 const core = require('@actions/core')
 const {exec} = require('@actions/exec')
-const {installElixir, installOTP} = require('./installer')
+const installer = require('./installer')
 const path = require('path')
 const semver = require('semver')
 const https = require('https')
 const {
-  fstat,
   promises: {readFile},
 } = require('fs')
 
-main().catch(err => {
+try {
+  main().catch((err) => {
+    core.setFailed(err.message)
+  })
+} catch (err) {
   core.setFailed(err.message)
-})
+}
 
 async function main() {
-  checkPlatform()
+  const maybePlatformIsOKError = installer.checkPlatform()
+  if (maybePlatformIsOKError !== 'ok') {
+    throw new Error(maybePlatformIsOKError)
+  }
 
   const osVersion = getRunnerOSVersion()
+  const otpBuildsListing = {
+    remote: `https://repo.hex.pm/builds/otp/${osVersion}/builds.txt`,
+  }
   const otpSpec = core.getInput('otp-version', {required: true})
-  const elixirSpec = core.getInput('elixir-version', {required: true})
-  const otpVersion = await getOtpVersion(otpSpec, osVersion)
-  const [elixirVersion, otpMajor] = await getElixirVersion(
-    elixirSpec,
-    otpVersion
-  )
-
-  let installHex = core.getInput('install-hex')
-  installHex = installHex == null ? 'true' : installHex
+  await installOTP(otpSpec, osVersion, otpBuildsListing)
+    .then(async (otpVersion) => {
+      const elixirBuildsListing = {
+        remote: 'https://repo.hex.pm/builds/elixir/builds.txt',
+      }
+      const elixirSpec = core.getInput('elixir-version', {required: false})
+      await maybeInstallElixir(elixirSpec, otpVersion, elixirBuildsListing)
+        .then(async (elixirInstalled) => {
+          if (elixirInstalled === true) {
+            const shouldMixRebar = core.getInput('install-rebar', {
+              required: false,
+            })
+            await mix(shouldMixRebar, 'rebar').catch((err) => {
+              throw err
+            })
+            const shouldMixHex = core.getInput('install-hex', {required: false})
+            await mix(shouldMixHex, 'hex').catch((err) => {
+              throw err
+            })
+          }
+        })
+        .catch((err) => {
+          throw err
+        })
```

Comment:
I don't know javascript very well, but is there any way we can avoid these `.catch((err) => { throw err })` everywhere? In the past we didn't use promises directly and instead only relied on the `async / await` syntax, can we still do that?

Reply:
We could, but, as stated before, we were not handling promise returns that eventually ended up hiding issues. Let me know what you prefer and I can do that. I also don't like this syntax very much, but it's what's working.

Reply:
I'll revert everything back to `async`, and potentially keep previously existing hidden bugs. Expect a couple of hours of implementation and testing, though, since I had a hard time getting to the point I'm at.

Reply:
This'll most likely also move us to an exception-based flow style of coding, which I found particularly hard to follow while adapting this. (still doing the move, though)

Reply:
I moved everything from `Promise` to `async` (include `Promise`-based code that previously existed also, for consistency). We know have many `throw`s where we didn't, but less `catch`, which seems to be what was bothering you. Lemme know if this approach is viable.